icu4c/source/i18n/inputext.cpp

```diff
@@ -64,7 +64,9 @@ void InputText::setDeclaredEncoding(const char* encoding, int32_t len)
         len += 1;     // to make place for the \0 at the end.
         uprv_free(fDeclaredEncoding);
         fDeclaredEncoding = NEW_ARRAY(char, len);
-        uprv_strncpy(fDeclaredEncoding, encoding, len);
+        if (fDeclaredEncoding != nullptr){
```

Comment:
We should try to allocate the new encoding string before deleting the old one.
The fix as is would leave the object in a bad state.